README.md
# markosamuli.pyenv

[![Ansible Quality Score](https://img.shields.io/ansible/quality/38342.svg)](https://galaxy.ansible.com/markosamuli/pyenv)
[![Ansible Role](https://img.shields.io/ansible/role/38342.svg)](https://galaxy.ansible.com/markosamuli/pyenv)
[![GitHub release](https://img.shields.io/github/release/markosamuli/ansible-pyenv.svg)](https://github.com/markosamuli/ansible-pyenv/releases)
[![License](https://img.shields.io/github/license/markosamuli/ansible-pyenv.svg)](https://github.com/markosamuli/ansible-pyenv/blob/master/LICENSE)

| Branch | Pipeline                                    |
| ------ | ------------------------------------------- |
| master | [![master][status-master]][pipeline-master] |

[pipeline-master]: https://github.com/markosamuli/ansible-pyenv/actions/workflows/pipeline.yml?query=branch%3Amaster
[status-master]: https://github.com/markosamuli/ansible-pyenv/workflows/Test%20and%20release/badge.svg?branch=master

Ansible role to install [pyenv] and [pyenv-virtualenv] on Ubuntu or macOS
development machines.

Optionally, [pyenv-virtualenvwrapper] can be installed and used for managing
environments.

Don't use this role on production servers as it supports installing pyenv only
under user home directory.

[pyenv]: https://github.com/pyenv/pyenv
[pyenv-virtualenv]: https://github.com/pyenv/pyenv-virtualenv
[pyenv-virtualenvwrapper]: https://github.com/pyenv/pyenv-virtualenvwrapper

## Install from Homebrew on macOS

The default method to install pyenv and plugins on macOS is to use Homebrew.

The role doesn't know how to migrate from existing Homebrew installs to
Git-based installations, so it will try to detect any existing installation
and keep using the previous method.

If you want to migrate, backup and delete your existing `~/.pyenv` directory
before running this role.

## Install from Homebrew on Linux

The role includes an experimental support for installing pyenv and plugins
with Homebrew on Linux.

The role doesn't install Homebrew on Linux itself and expects it to be installed
in the default `/home/linuxbrew/.linuxbrew` location.

Installing Python versions with pyenv on Linux when Homebrew installation exists
has some known issues:

- readline extension was not compiled, installed pyenv by Linuxbrew on
  Ubuntu 16 [#1479][pyenv-1479]

[pyenv-1479]: https://github.com/pyenv/pyenv/issues/1479

## Installed Python versions

This role installs [Python][python] versions defined in `pyenv_python_versions`
variable.

To set global version, set `pyenv_global` variable to the desired version(s).

```yaml
pyenv_global: "{{ pyenv_python37_version }} system"
```

This is configured to use latest Python 2 and Python 3 versions and the
system version as default.

[python]: https://www.python.org

## Changes to shell config files

This role creates config file in `~/.pyenv/.pyenvrc` that is loaded in
`.bashrc` and `.zshrc` files.

Code completion is loaded by default.

If you're managing your shell scripts `.dotfiles` or are using a framework, you
should set `pyenv_init_shell` to `false` and update these files yourself to keep
them clean.

Reference `.bashrc` configuration:

```bash
if [ -e "$HOME/.pyenv/.pyenvrc" ]; then
  source $HOME/.pyenv/.pyenvrc
  if [ -e "$HOME/.pyenv/completions/pyenv.bash" ]; then
    source $HOME/.pyenv/completions/pyenv.bash
  elif [ -e "/usr/local/opt/pyenv/completions/pyenv.bash" ]; then
    source /usr/local/opt/pyenv/completions/pyenv.bash
  fi
fi
```

Reference `.zshrc` configuration:

```zsh
if [ -e "$HOME/.pyenv/.pyenvrc" ]; then
  source $HOME/.pyenv/.pyenvrc
  if [ -e "$HOME/.pyenv/completions/pyenv.zsh" ]; then
    source $HOME/.pyenv/completions/pyenv.zsh
  elif [ -e "/usr/local/opt/pyenv/completions/pyenv.zsh" ]; then
    source /usr/local/opt/pyenv/completions/pyenv.zsh
  fi
fi
```

## Role Variables

Path to `~/.pyenv` is based on environment variables:

```yaml
pyenv_home: "{{ ansible_env.HOME }}"
pyenv_root: "{{ ansible_env.HOME }}/.pyenv"
```

Update `.bashrc` and `.zshrc` files in user home directory:

```yaml
pyenv_init_shell: true
```

Versions to install:

```yaml
pyenv_version: "v1.2.13"
pyenv_virtualenv_version: "v1.1.5"
pyenv_virtualenvwrapper_version: "v20140609"
```

Latest Python 3.7 and Python 3.8 versions:

```yaml
pyenv_python37_version: "3.7.6"
pyenv_python38_version: "3.8.1"
```

Python 2 and Python 3 versions are installed by default:

```yaml
pyenv_python_versions:
  - "{{ pyenv_python37_version }}"
  - "{{ pyenv_python38_version }}"
```

Set global version to Python 3.7 with `system` fallback:

```yaml
pyenv_global: "{{ pyenv_python37_version }} system"
```

Install virtualenvwrapper plugin:

```yaml
pyenv_virtualenvwrapper: false
pyenv_virtualenvwrapper_home: "{{ ansible_env.HOME }}/.virtualenvs"
```

Install using Homebrew package manager on macOS:

```yaml
pyenv_install_from_package_manager: true
```

Detect existing installation method and use that:

```yaml
pyenv_detect_existing_install: true
```

Install using Homebrew on Linux:

```yaml
pyenv_homebrew_on_linux: true
```

## Example Playbook

```yaml
- hosts: localhost
  connection: local
  become: false
  roles:
    - role: markosamuli.pyenv
```

## Updating versions

Run the following scripts to get latest releases from GitHub and update them in
role defaults.

Update [pyenv] release:

```bash
./update-release pyenv
```

Update [pyenv-virtualenv] release:

```bash
./update-release pyenv-virtualenv
```

Update default [Python] 3.7 version:

```bash
./update-python python37
```

Update default [Python] 3.8 version:

```bash
./update-python python38
```

Update all versions:

```bash
make update
```

## Coding style

Install pre-commit hooks and validate coding style:

```bash
make lint
```

## Run tests

Run tests in Ubuntu and Debian using Docker:

```bash
make test
```

## Acknowledgements

Use of `.pyenvrc` file and parts used for installing python version taken from
[avanov.pyenv](https://github.com/avanov/ansible-galaxy-pyenv) role.

## Development

Install development dependencies in a local virtualenv:

```bash
make setup
```

Install [pre-commit] hooks:

```bash
make install-git-hooks
```

[pre-commit]: https://pre-commit.com/

## Changes

- [CHANGELOG.md](CHANGELOG.md)

## License

- [MIT](LICENSE)

## Author Information

- [@markosamuli](https://github.com/markosamuli)


CHANGELOG.md
# Changelog

Only the latest major version is maintained.

## [6.2.0] - 2022-04-25

Support for Ubuntu 22.04 LTS and Arch Linux. I don't have test environments for
either of these so testing is limited to the Docker environments only.

[6.2.0]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v6.2.0

### Fixed

- Change how tasks are included to address issues with missing homebrew action
  on Linux. ([#47][issue-47])

[issue-47]: https://github.com/markosamuli/ansible-pyenv/issues/47

### Changed

- Added tests for Ubuntu 22.04 LTS.

  Compiling Python 3.10 with Homebrew fails on Ubuntu 22.04 LTS so Homebrew
  tests are not included.

- Added configuration and tests for Arch Linux.

  Do not test Homebrew on Arch Linux as Homebrew install itself fails due to
  incorrect Ruby version.

- Default to `python3-openssl` in Ubuntu and Debian. ([#54][issue-54])

  This should work as the new default since Python 2 is EOL.

[issue-54]: https://github.com/markosamuli/ansible-pyenv/issues/54

### Development changes

- Test macOS install on GitHub Actions.
- Remove `develop` branch and pipelines.
- Support for running cross platform test images on Apple M1 devices.
- Use [`robertdebock/galaxy-action`][galaxy-action] GitHub Action to publish
  tagged releases to Ansible Galaxy once tests have been completed.

[galaxy-action]: https://github.com/robertdebock/galaxy-action

## [6.1.0] - 2022-04-15

Fix breaking changes introduced in pyenv 2.0.0.

[6.1.0]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v6.1.0

### Fixed

- Move PATH configuration to `.zprofile`, `.bash_profile` and `.profile` files
- Add `community.general` into collections to make `homebrew` available on new
  Ansible versions

## [6.0.0] - 2022-04-15

Use pyenv 2.2.5, Python 3.10 and add support for installing with Homebrew on
Apple M1 chips.

[6.0.0]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v6.0.0

### Breaking changes

- Update pyenv from 1.2.27 to 2.2.5
- Use Python 3.10 as the default version
- Do not install Python 3.8 or Python 3.9 by default
- Require Ansible 2.10

### Changed

- Add Python 3.10.3
- Update Python 3.7.9 to 3.7.13
- Update Python 3.8.9 to 3.8.13
- Update Python 3.9.4 to 3.9.11
- Run tests on the CI with Python 3.10
- Update pre-commit hooks to the latest versions

### Fixed

- Added build requirements for Debian bullseye
- Python version update script was not sorting versions correctly
- Update Homebrew installation path to `/opt/homebrew` on Apple M1 chips
- Do not try to run tests with Homebrew on Linux on ARM
- Upgrade jinja2 to 3.1.1
- Upgrade gitpython to 3.1.27
- shfmt install on CI pipelines

## [5.0.0] - 2021-05-06

New release tested with remote install on macOS and Ubuntu Linux.

[5.0.0]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v5.0.0

### Breaking changes

- Use Python 3.8 as the default version
- Do not install Python 3.7 automatically

### Fixed

- Do not print `docker` path in run-tests.sh
- Failing `shellcheck` pre-commit hook

### Changed

Use `pyenv` binary directly instead of loading the `.pyenvrc` file as suggested
by [@KentBrockman](https://github.com/KentBrockman) in [PR #39][pr39]. This will
allow installation of the role using remote SSH connections without the need
for interactive shell.

Remove links to Travis CI as the [tests are no longer working][travis] for
open source projects.

Remove support for old operating systems:

- Use Ubuntu 20.04 LTS as the default environment on Travis
- Remove Ubuntu 16.04 LTS tests
- Remove macOS High Sierra tests
- Remove Debian stretch tests

Improve local development setup:

- Manage development dependencies with [pip-tools]
- Use local virtualenv when running tasks in the Makefile
- Update bandit from 1.6.2 to 1.7.0

Update development dependencies with vulnerabilities:

- Update pylint from 2.5.2 to 2.7.0
- Bump jinja2 from 2.11.2 to 2.11.3

Update pyenv and Python versions:

- Install Python 3.9.4
- Update pyenv to 1.2.27
- Update default python versions to 3.7.8 and 3.8.9

[pip-tools]: https://github.com/jazzband/pip-tools
[travis]: https://github.com/markosamuli/ansible-pyenv/issues/45
[pr39]: https://github.com/markosamuli/ansible-pyenv/pull/39

## [4.0.2] - 2020-09-11

### Fixed

- Update apt cache when installing Linux dependencies to fix issue with build
  requirements failing to install on fresh installations [#37][issue37]
  ([@papierukartka][papierukartka])
- Run GitHub [Test](.github/workflows/test.yml) and
  [Code Quality](.github/workflows/code-quality.yml) workflows on `bugfix/*`
  branches and pull requests targeting the `develop` or `master` branches

[issue37]: https://github.com/markosamuli/ansible-pyenv/issues/37
[papierukartka]: https://github.com/papierukartka
[4.0.2]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v4.0.2

## [4.0.1] - 2020-09-05

This release addresses installation issues and adds additional test environments
using Docker images.

[4.0.1]: https://github.com/markosamuli/ansible-pyenv/releases/tag/v4.0.1

### Fixed

- Run `brew into` in Ansible `shell` module instead of `command` to use the
  existing shell environment.
- Do not use Homebrew OpenSSL on Debian stretch or Ubuntu xenial to fix Python
  3.7 and 3.8 install failing due to OpenSSL not being found.
- Debian: Install `python3-openssl` instead of `python-openssl` on Debian
  buster.
- Add `black` compatible `flake8` configuration.

### Changed

Improved `run-tests.sh` script and error handling if any of the tests fail.

- Add `--debug` option to the `run-tests.sh` script that will start a
  container with an interactive shell to allow ad-hoc commands to be
  executed inside the container.

Improve the `update_test_images.py` script to allow additional options to
be provided from the command line.

- Use `click` in the `update_test_images.py` script to allow additional options
  to be provided.
- Add `--list-only` option to return the Dockerfile build target paths

tests/utils.sh
#!/usr/bin/env bash

TESTS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PROJECT_ROOT=$(dirname "$TESTS_DIR")
export PROJECT_ROOT

# Paths in which Ansible will search for Roles
ANSIBLE_ROLES_PATH=$(dirname "$PROJECT_ROOT")
export ANSIBLE_ROLES_PATH

msg() {
    local message="$*"
    if [ -n "${TEST_ENV}" ]; then
        message="[${TEST_ENV}] ${message}"
    fi
    echo "${message}"
}

syntax_check() {
    echo "*** Check syntax"
    if ansible-playbook tests/test.yml -i tests/inventory --syntax-check; then
        msg "Syntax check: pass"
    else
        msg "Syntax check: fail"
        return 1
    fi
}

run_tests() {
    local ansible_vars=("$@")
    local extra_vars
    extra_vars=$(printf ",%s" "${ansible_vars[@]}")
    extra_vars="{${extra_vars:1}}"

    msg "*** Run Ansible playbook"
    if run_playbook "${extra_vars}"; then
        msg "Playbook run: pass"
    else
        msg "Playbook run: fail"
        return 1
    fi

    msg "*** Idempotence test"
    if run_playbook "${extra_vars}" | grep -q 'changed=0.*failed=0'; then
        msg "Idempotence test: pass"
    else
        msg "Idempotence test: fail"
        return 1
    fi
}

run_playbook() {
    local extra_vars="$1"
    ansible-playbook tests/test.yml -i tests/inventory --connection=local \
        -e "${extra_vars}" -v
}


tests/update.sh
#!/usr/bin/env bash

TESTS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PROJECT_ROOT=$(dirname "$TESTS_DIR")

TEST_USER="test"
REPOSITORY="$(basename "$PROJECT_ROOT")"
ANSIBLE_VERSION="<2.9.0,!=2.8.6"

# Generate Dockerfile from Dockerfile.template
update_dockerfile() {
    local docker_image=$1
    local docker_tag=$2
    echo "*** Updating ${docker_tag}/Dockerfile"
    mkdir -p "${docker_tag}"
    if test "$(uname)" = "Darwin"; then
        sed -E \
            -e 's!%%FROM%%!'"${docker_image}:${docker_tag}"'!g' \
            -e 's!%%USER%%!'"${TEST_USER}"'!g' \
            -e 's!%%REPOSITORY%%!'"${REPOSITORY}"'!g' \
            -e 's~%%ANSIBLE_VERSION%%~'"${ANSIBLE_VERSION}"'~g' \
            "Dockerfile.template" >"${docker_tag}/Dockerfile"
    else
        sed -r \
            -e 's!%%FROM%%!'"${docker_image}:${docker_tag}"'!g' \
            -e 's!%%USER%%!'"${TEST_USER}"'!g' \
            -e 's!%%REPOSITORY%%!'"${REPOSITORY}"'!g' \
            -e 's~%%ANSIBLE_VERSION%%~'"${ANSIBLE_VERSION}"'~g' \
            "Dockerfile.template" >"${docker_tag}/Dockerfile"
    fi
}

set -euo pipefail

cd "${TESTS_DIR}"

ubuntu_releases=(xenial bionic)
for tag in "${ubuntu_releases[@]}"; do
    update_dockerfile "ubuntu" "${tag}"
done

debian_releases=(stretch buster)
for tag in "${debian_releases[@]}"; do
    update_dockerfile "debian" "${tag}"
done


tests/run-tests.sh
#!/usr/bin/env bash

TESTS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PROJECT_ROOT=$(dirname "$TESTS_DIR")

ROLE_NAME="$(basename "$PROJECT_ROOT")"
TEST_HOME=/home/test

if [[ $(uname -m) == arm64 ]]; then
    USE_BUILDKIT=${USE_BUILDKIT:-true}
else
    USE_BUILDKIT=${USE_BUILDKIT:-false}
fi

IMAGES_DIR="images"

docker_run_opts=()

debug="false"
test_with_git="true"
test_with_homebrew="true"

error() {
    echo "$@" 1>&2
}

detect_wsl() {
    is_wsl=0
    if [[ -d "/run/WSL" ]]; then
        echo "*** Windows Subsystem for Linux (WSL2) detected"
        is_wsl=1
        wsl_version='WSL2'
        return
    fi
    if [ ! -e /proc/version ]; then
        return
    fi
    if grep -q Microsoft /proc/version; then
        echo "*** Windows Subsystem for Linux (WSL1) detected"
        is_wsl=1
        wsl_version='WSL1'
        return
    fi
}

setup_mount_root() {
    if [ "${is_wsl}" == "1" ] && [ "${wsl_version}" == "WSL1" ]; then
        MOUNT_ROOT="$(pwd -P | sed 's~/mnt/c/~c:/~')"
    else
        MOUNT_ROOT="$(pwd -P)"
    fi
}

setup_docker_opts() {
    if [ -z "$CI" ]; then
        docker_run_opts+=("-it")
    fi
}

finish() {
    local containers=""
    containers=$(docker ps -q --filter=name="${ROLE_NAME}")
    if [ -n "${containers}" ]; then
        echo "*** Stop all test containers"
        # shellcheck disable=SC2086
        docker stop ${containers}
    fi
}

stop() {
    local image=$1
    local container_name="${ROLE_NAME}-${image}-tests"
    echo "*** Stop container with ${image}"
    docker stop "${container_name}"
}

build() {
    local image=$1
    local image_name="${ROLE_NAME}-${image}"
    local image_dir="${TESTS_DIR}/${IMAGES_DIR}/${image}"
    local dockerfile="${image_dir}/Dockerfile"
    local base_image
    base_image=$(grep "^FROM" "${dockerfile}" | sed 's/FROM //')
    echo "*** Pull base image ${base_image}"
    docker pull "${base_image}"
    echo "*** Build image ${image}"
    docker build -t "${image_name}" "${image_dir}"
}

build_using_buildkit() {
    local image=$1
    local platform=$2
    local image_name="${ROLE_NAME}-${image}"
    local image_dir="${TESTS_DIR}/${IMAGES_DIR}/${image}"
    local dockerfile="${image_dir}/Dockerfile"
    echo "*** Build image ${image}"
    docker buildx build --platform "${platform}" -t "${image_name}" "${image_dir}"
}

start() {
    local image=$1
    local platform=$2
    local image_name="${ROLE_NAME}-${image}"
    local container_name="${ROLE_NAME}-${image}-tests"
    local start_opts=()
    if [ -n "${platform}" ]; then
        start_opts+=("--platform")
        start_opts+=("${platform}")
    fi
    echo "*** Start container with image ${image}"
    docker run --rm -d \
        "${docker_run_opts[@]}" \
        "${start_opts[@]}" \
        -e "TEST_ENV=${image}" \
        -v "${MOUNT_ROOT}:${TEST_HOME}/${ROLE_NAME}" \
        --name "${container_name}" \
        "${image_name}" || {
        error "failed to start container"
        exit 1
    }
}

run_tests() {
    local image=$1
    local platform=$2
    if [[ $image == *"homebrew"* ]]; then
        if [ "${test_with_homebrew}" == "true" ]; then
            run_tests_with_homebrew "$image" "$platform" || {
                error "failed Homebrew installation in ${image}"
                return 1
            }
        fi
    else
        if [ "${test_with_git}" == "true" ]; then
            run_tests_with_git "$image" "$platform" || {
                error "failed Git installation in ${image}"
                return 1
            }
        fi
    fi
}

run_tests_with_git() {
    local image=$1
    local platform=$2
    local test_scripts=(
        "test_syntax.sh"
        "test_install.sh"
    )
    local failed=0
    echo "*** Run tests installing with Git install"
    for test_script in "${test_scripts[@]}"; do
        start "${image}" "${platform}"
        run_test_script "${image}" "${test_script}" || failed=1
        stop "${image}"
        # Give Docker time to clean up
        sleep 1
    done
    if [ "${failed}" -gt "0" ]; then
        return "${failed}"
    fi
}

run_tests_with_homebrew() {
    local image=$1
    local platform=$2
    local test_scripts=(
        "test_syntax.sh"
        "test_install_with_homebrew.sh"
    )
    local failed=0
    echo "*** Run tests installing with Homebrew install"
    for test_script in "${test_scripts[@]}"; do
        start "${image}" "${platform}"
        run_test_script "${image}" "${test_script}" || failed=1
        stop "${image}"
        # Give Docker time to clean up
        sleep 1
    done
    if [ "${failed}" -gt "0" ]; then
        return "${failed}"
    fi
}

run_test_script() {
    local image=$1
    local test_script=$2
    local container_name="${ROLE_NAME}-${image}-tests"
    echo "*** Run tests with ${test_script} in ${image}"
    docker exec \
        "${docker_run_opts[@]}" \
        --user test \
        "${container_name}" \
        bash -ilc "${TEST_HOME}/${ROLE_NAME}/tests/${test_script}" || {
        error "${test_script} on ${image} failed"
        return 1
    }
}

debug_image() {
    local image=$1
    local container_name="${ROLE_NAME}-${image}-tests"
    start "${image}"
    docker exec \
        "${docker_run_opts[@]}" \
        --user test \
        "${container_name}" \
        bash -il
}

if ! command -v docker >/dev/null; then
    error "docker not found"
    exit 1
fi

# Parse command line arguments
POSITIONAL=()
while [[ $# -gt 0 ]]; do
    key="$1"
    case $key in
        --debug)
            debug="true"
            shift
            ;;
        --with-git)
            test_with_git="true"
            shift
            ;;
        --without-git)
            test_with_git="false"
            shift
            ;;
        --with-homebrew)
            test_with_homebrew="true"
            shift
            ;;
        --without-homebrew)
            test_with_homebrew="false"
            shift
            ;;
        *)                     # unknown option
            POSITIONAL+=("$1") # save it in an array for later
            shift              # past argument
            ;;
    esac
done
set -- "${POSITIONAL[@]}" # restore positional parameters

trap finish EXIT

detect_wsl

setup_docker_opts

cd "${TESTS_DIR}"

images=("$@")
if [ ${#images[@]} -eq 0 ]; then
    images=(images/*/Dockerfile)
fi
images=("${images[@]/${PROJECT_ROOT}\//}")
images=("${images[@]/tests\//}")
images=("${images[@]/images\//}")
images=("${images[@]/\/Dockerfile/}")

platforms=(
    "linux/amd64"
)

cd "${PROJECT_ROOT}"

setup_mount_root

set -e

for i in "${images[@]}"; do
    if [ "${USE_BUILDKIT}" == "true" ]; then
        for p in "${platforms[@]}"; do
            build_using_buildkit "$i" "$p" || {
                error "failed to build $i for platform $p"
                exit 1
            }
        done
    else
        build "$i" || {
            error "failed to build $i"
            exit 1
        }
    fi
done

if [ "${debug}" == "true" ]; then
    if [ ${#images[@]} -gt 1 ]; then
        error "limit to one Docker image to debug"
        exit 1
    fi
    debug_image "${images[0]}"
else
    for i in "${images[@]}"; do
        if [ "${USE_BUILDKIT}" == "true" ]; then
            for p in "${platforms[@]}"; do
                run_tests "$i" "$p" || {
                    error "failed tests $i on platform $p"
                    exit 1
                }
            done
        else
            run_tests "$i" || {
                error "failed tests in $i"
                exit 1
            }
        fi
    done
fi


scripts/utils.sh
#!/usr/bin/env bash

# Print error into STDERR
error() {
    echo "$@" 1>&2
}

# Get latest release for a GitHub repository
get_latest_release() {
    local repository=$1
    local url="https://api.github.com/repos/${repository}/releases/latest"
    if [ -n "$GITHUB_OAUTH_TOKEN" ]; then
        url="${url}?access_token=${GITHUB_OAUTH_TOKEN}"
    fi
    if test "$(uname)" = "Darwin"; then
        curl --silent "${url}" |
            grep '"tag_name":' |
            sed -E 's/.*"([^"]+)".*/\1/'
    else
        curl --silent "${url}" |
            grep '"tag_name":' |
            sed -r 's/.*"([^"]+)".*/\1/'
    fi
}

# Get latest tag from a GitHub repository
get_latest_tag() {
    local repository=$1
    local url="https://api.github.com/repos/${repository}/tags?per_page=1"
    if [ -n "$GITHUB_OAUTH_TOKEN" ]; then
        url="${url}&access_token=${GITHUB_OAUTH_TOKEN}"
    fi
    if test "$(uname)" = "Darwin"; then
        curl --silent "${url}" |
            grep '"name":' |
            sed -E 's/.*"([^"]+)".*/\1/'
    else
        curl --silent "${url}" |
            grep '"name":' |
            sed -r 's/.*"([^"]+)".*/\1/'
    fi
}

# Update Ansible variable in defaults/main.yml
update_default_variable() {
    local key=$1
    local value=$2
    local file=defaults/main.yml
    update_variable_file "${file}" "${key}" "${value}"
}

# Update Ansible variable in vars/os/ansible_os.yml
update_os_variable() {
    local os=$1
    local key=$2
    local value=$3

    if [ -z "${os}" ]; then
        error "OS name missing"
        exit 1
    fi
    if [ -z "${key}" ]; then
        error "Variable name missing"
        exit 1
    fi
    if [ -z "${value}" ]; then
        error "Variable value missing"
        exit 1
    fi

    local ansible_os
    if [ "${os}" == "darwin" ] || [ "${os}" == "Darwin" ]; then
        ansible_os="Darwin"
    elif [ "${os}" == "debian" ] || [ "${os}" == "Debian" ]; then
        ansible_os="Debian"
    fi
    if [ -z "${ansible_os}" ]; then
        error "Unsupported OS: ${os}"
        exit 1
    fi

    local file="vars/os/${ansible_os}.yml"
    update_variable_file "${file}" "${key}" "${value}"
}

# Update variable in Ansible file
update_variable_file() {
    local file=$1
    local key=$2
    local value=$3
    if [ ! -e "${file}" ]; then
        error "${file} not found"
        exit 1
    fi
    if test "$(uname)" = "Darwin"; then
        sed -i.save -E "s/^($key):.*$/\1: \"${value}\"/" \
            "${file}"
    else
        sed -i.save -r "s/^($key):.*$/\1: \"${value}\"/" \
            "${file}"
    fi
    rm "${file}.save"
}

# Check cURL is installed
check_curl() {
    command -v curl >/dev/null || {
        error "cURL is not installed"
        exit 1
    }
}

# Check jq is installed
check_jq() {
    command -v jq >/dev/null || {
        error "jq not installed"
        exit 1
    }
}


scripts/update-release.sh
#!/usr/bin/env bash

SCRIPTS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PROJECT_ROOT=$(dirname "$SCRIPTS_DIR")

# shellcheck source=scripts/utils.sh
source "${SCRIPTS_DIR}/utils.sh"

# Update pyenv version
update_pyenv_version() {
    local version=$1
    update_default_variable "pyenv_version" "${version}"
}

# Update pyenv-virtualenv version
update_pyenv_virtualenv_version() {
    local version=$1
    update_default_variable "pyenv_virtualenv_version" "${version}"
}

# Get latest pyenv version
latest_pyenv_version() {
    local latest_release=""
    local errmsg="Failed to get latest pyenv/pyenv release"
    latest_release=$(get_latest_release pyenv/pyenv)
    if [ -z "${latest_release}" ]; then
        error "${errmsg}"
        exit 1
    fi
    echo "${latest_release}"
}

# Get latest pyenv-virtualenv version
latest_pyenv_virtualenv_version() {
    local latest_release=""
    local errmsg="Failed to get latest pyenv/pyenv-virtualenv release"
    latest_release=$(get_latest_release pyenv/pyenv-virtualenv)
    if [ -z "${latest_release}" ]; then
        error "${errmsg}"
        exit 1
    fi
    echo "${latest_release}"
}

# Update all versions
update_versions() {
    local repository=$1
    local version
    if [ -z "${repository}" ]; then
        error "Repository not defined"
        exit 1
    fi
    if [ "${repository}" == "pyenv" ]; then
        version=$(latest_pyenv_version)
        echo "Latest pyenv release is ${version}"
        update_pyenv_version "${version}"
    elif [ "${repository}" == "pyenv-virtualenv" ]; then
        version=$(latest_pyenv_virtualenv_version)
        echo "Latest pyenv-virtualenv release is ${version}"
        update_pyenv_virtualenv_version "${version}"
    else
        error "Unknown repository: ${repository}"
        exit 1
    fi
}

set -e

cd "${PROJECT_ROOT}"

check_curl
update_versions "$1"


scripts/update-python.sh
#!/usr/bin/env bash

SCRIPTS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PROJECT_ROOT=$(dirname "$SCRIPTS_DIR")

# shellcheck source=scripts/utils.sh
source "${SCRIPTS_DIR}/utils.sh"

# Get latest stable tag for a GitHub repository
get_latest_cpython_version() {
    local target_version=$1
    local repository=python/cpython
    local url="https://api.github.com/repos/${repository}/git/refs/tags"
    if [ -n "$GITHUB_OAUTH_TOKEN" ]; then
        url="${url}?access_token=${GITHUB_OAUTH_TOKEN}"
    fi
    # @formatter:off
    curl --silent "${url}" |
        jq --arg version_tag "${target_version}" \
            -r 'map(select(.object.type == "tag"))
            | map(select(.ref | test("refs/tags/" + $version_tag)))
            | map(select(.ref | test("rc|0[ab][0-9]") | not))
            | map(.ref)
            | map(capture("refs/tags/v(?<version>.*)"))
            | sort_by(.version | split(".") | map(tonumber))
            | last | .version'
    # @formatter:on
}

update_versions() {
    local target=$1
    if [ -z "${target}" ]; then
        error "Target version not defined"
        exit 1
    fi
    if [ "${target}" == "python2" ]; then
        error "Python 2.7 is no longer supported"
        exit 1
    elif [ "${target}" == "python3" ]; then
        error "Use python37, python38 or python310 as the target version"
        exit 1
    elif [ "${target}" == "python37" ]; then
        update_python "python37" "3.7"
    elif [ "${target}" == "python38" ]; then
        update_python "python38" "3.8"
    elif [ "${target}" == "python39" ]; then
        update_python "python39" "3.9"
    elif [ "${target}" == "python310" ]; then
        update_python "python310" "3.10"
    else
        error "Unknown target version: ${target}"
        exit 1
    fi
}

update_python() {
    local target=$1
    local target_version=$2
    local latest_version
    latest_version=$(get_latest_cpython_version "v${target_version}")
    echo "Latest Python ${target_version} release is ${latest_version}"
    update_default_variable "pyenv_${target}_version" "${latest_version}"
}

set -e

cd "${PROJECT_ROOT}"

check_curl
check_jq

update_versions "$1"
